go/client/logclient_test.go

```diff
@@ -104,3 +109,80 @@ func TestGetSTHWorks(t *testing.T) {
 		t.Fatalf("Invalid TreeHeadSignature.Signature: expected %v, got %v", sth.TreeHeadSignature.Signature, expectedDS.Signature)
 	}
 }
+
+func TestAddChainWithContext(t *testing.T) {
+	retryAfter := 0
+	currentFailures := 0
+	failuresBeforeSuccess := 0
+	hs := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if failuresBeforeSuccess > 0 && currentFailures < failuresBeforeSuccess {
+			currentFailures++
+			if retryAfter > 0 {
+				w.Header().Add("Retry-After", strconv.Itoa(retryAfter))
+				w.WriteHeader(503)
+				return
+			}
+			w.WriteHeader(408)
+			return
+		}
+		_, err := w.Write([]byte(`{"sct_version":0,"id":"KHYaGJAn++880NYaAY12sFBXKcenQRvMvfYE9F1CYVM=","timestamp":1337,"extensions":"","signature":"BAMARjBEAiAIc21J5ZbdKZHw5wLxCP+MhBEsV5+nfvGyakOIv6FOvAIgWYMZb6Pw///uiNM7QTg2Of1OqmK1GbeGuEl9VJN8v8c="}`))
+		if err != nil {
+			return
+		}
+	}))
+	defer hs.Close()
+
+	certBytes, err := base64.StdEncoding.DecodeString(SubmissionCertB64)
+	if err != nil {
+		t.Fatalf("Failed to decode chain array B64: %s", err)
+	}
+	chain := []ct.ASN1Cert{certBytes}
+
+	c := New(hs.URL)
+	leeway := time.Millisecond * 100
+	instant := time.Millisecond
+	fiveSeconds := time.Second * 5
+
+	testCases := []struct {
+		deadlineLength        int
+		expected              time.Duration
+		retryAfter            int
+		failuresBeforeSuccess int
+		success               bool
+	}{
+		{-1, instant, 0, 0, true},
+		{6, fiveSeconds, 5, 1, true},
+		{5, fiveSeconds, 10, 1, false},
+		{10, fiveSeconds, 1, 5, true},
+		{1, instant * 10, 0, 10, true},
+	}
+
+	for _, tc := range testCases {
+		var deadline context.Context
+		if tc.deadlineLength >= 0 {
+			deadline, _ = context.WithDeadline(context.Background(), time.Now().Add(time.Duration(tc.deadlineLength)*time.Second))
+		}
+		retryAfter = tc.retryAfter
+		failuresBeforeSuccess = tc.failuresBeforeSuccess
+		currentFailures = 0
+
+		started := time.Now()
+		sct, err := c.AddChainWithContext(deadline, chain)
+		took := time.Since(started)
+		margin := leeway
+		if margin > tc.expected {
+			margin = 0
+		}
+		if took > tc.expected+leeway || took < tc.expected-leeway {
+			t.Fatalf("Submission took an unexpected length of time: %s, expected ~%s", took, tc.expected)
+		}
+		if tc.success && err != nil {
+			t.Fatalf("Failed to submit chain: %s", err)
+		} else if err == nil && !tc.success {
```

Comment:
Perhaps consider swapping the order of the comparison args here to match the earlier if for consistency?